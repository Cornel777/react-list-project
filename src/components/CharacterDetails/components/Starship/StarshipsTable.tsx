import { TableContainer, Paper, Table, TableHead, TableBody } from "@mui/material";
import { FunctionComponent } from "react";
import { StyledTableRow, StyledTableCell } from "../styled/styledTable";
import Starship from "./Starship";

interface StarshipsTableProps {
  data: {
    starships: string[];
  }
}

const StarshipsTable: FunctionComponent<StarshipsTableProps> = ({ data }) => {

  if (!data.starships.length) {
    return <p>No starships flown</p>
  }

  return (
    <TableContainer component={Paper} style={{ maxWidth: '400px' }}>
        <Table size="medium" aria-label="Starships table">
         <TableHead>
            <StyledTableRow>
              <StyledTableCell>Starship name</StyledTableCell>
               <StyledTableCell>Model</StyledTableCell>
            </StyledTableRow>
          </TableHead>
          <TableBody>
            {data.starships.map((starshipUrl: string) => {
              return <Starship starshipUrl={starshipUrl} />
            })}
          </TableBody>
        </Table>
      </TableContainer>
  )
}

export default StarshipsTable;