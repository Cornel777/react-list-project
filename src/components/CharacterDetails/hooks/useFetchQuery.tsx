import { useQuery, keepPreviousData } from "@tanstack/react-query";
import fetchSwapi from "../../../services/api/fetchSwapi";

export interface HookReturn<T> {
  isFetching: boolean;
  isPlaceholderData: boolean;
  isPending: boolean;
  isError: boolean;
  error: Error | null;
  data: T | undefined;
}


const useFetchQuery = <T, >(queryKey: string, path: string, id: string | undefined): HookReturn<T> => {

  const { isFetching, isPlaceholderData, isPending, isError, error, data } = useQuery<T>({
    queryKey: [queryKey, id],
    queryFn: () => fetchSwapi(path),
    staleTime: 5000,
    placeholderData: keepPreviousData,
  })

    return {
      isFetching,
      isPlaceholderData,
      isPending,
      isError,
      error,
      data,
    }
}

export default useFetchQuery;