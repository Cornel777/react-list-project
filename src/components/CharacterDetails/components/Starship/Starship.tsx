import { FunctionComponent } from "react";
import { StyledTableRow, StyledTableCell } from "../styled/styledTable";
import { type Starship } from "../../../../types/Starship";
import useFetchQuery from "../../hooks/useFetchQuery";


interface StarshipProps {
  starshipUrl: string;
}

const Starship: FunctionComponent<StarshipProps> = ({ starshipUrl }) => {
  const starshipUrlList = starshipUrl.split('/').filter((element: string) => element.length > 0);
  const starshipId = starshipUrlList.pop();

  const { isPending, isError, error, data} = useFetchQuery<Starship>('starship', `starships/${starshipId}/`, starshipId);

  if (isPending) {
    return <p>Loading...</p>
  }
  if (isError) {
    return <p>{`Failed to load starship ${error && error.message}`}</p>
  }
  if (!data) {
    return null;
  }

  return (
    <StyledTableRow key={starshipId}>
      <StyledTableCell>{data.name }</StyledTableCell>
      <StyledTableCell>{data.model}</StyledTableCell>
    </StyledTableRow>
  )
}

export default Starship;