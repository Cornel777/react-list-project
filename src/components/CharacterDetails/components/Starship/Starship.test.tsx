import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import useFetchQuery from '../../hooks/useFetchQuery';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import Starship from './Starship';

vi.mock('../../hooks/useFetchQuery');

const mockStarshipData = {
  name: 'X-wing',
  model: 'T-65 X-wing',
  manufacturer: 'Incom Corporation',
  url: 'https://swapi.dev/api/starships/12/',
  passengers: '0',
  cargo_capacity: '110',
  consumables: '1 week',
  cost_in_credits: '149999',
  crew: '1',
  hyperdrive_rating: '1.0',
  length: '12.5',
  max_atmosphering_speed: '1050',
  starship_class: 'Starfighter'
};

describe('Starship', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false
      }
    }
  });

  beforeEach(() => {
    vi.clearAllMocks();
    vi.mocked(useFetchQuery).mockReturnValue({
      isPending: false,
      isError: false,
      isFetching: false,
      isPlaceholderData: false,
      error: null,
      data: mockStarshipData
    });
  });

  function renderWithClient(ui: React.ReactElement) {
    const testQueryClient = queryClient;
    const { rerender, ...result } = render(
        <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>
    )
    return {
        ...result,
        rerender: (rerenderUi: React.ReactElement) =>
            rerender(
                <QueryClientProvider client={testQueryClient}>{rerenderUi}</QueryClientProvider>
            ),
    }
  }

  it('matches the snapshot', () => {
    const { asFragment } = renderWithClient(<Starship starshipUrl='https://swapi.dev/api/starships/12/' />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders homeworld details', () => {
    renderWithClient(<Starship starshipUrl='https://swapi.dev/api/starships/12/' />)
    expect(screen.getByText('X-wing')).toBeInTheDocument();
    expect(useFetchQuery).toHaveBeenCalledWith('starship', 'starships/12/', '12');
  });
});