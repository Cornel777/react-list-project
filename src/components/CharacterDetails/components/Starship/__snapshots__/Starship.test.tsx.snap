// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Starship > matches the snapshot 1`] = `
<DocumentFragment>
  X-wingT-65 X-wing
</DocumentFragment>
`;
